enhancements/installer/aws-custom-dns.md

```diff
@@ -0,0 +1,488 @@
+ConfigResource(CR) with the configuration required to configure DNS for these
+services.
+
+7. The MCO will watch the DNSRecord CR for the DNS configuration mentioned in
+step 6. MCO will update the newly added `IngressDNSConfig` field within the
+`AWSPlatformStatus` with the DNS configuartion for *.apps records by reading
+the DNSRecord CR.
+
+8. The runtimecfg utility will monitor the `Infrastructure` CR for changes to
+`IngressDNSConfig` of `AWSPlatformStatus` to update the CoreDNS pod's
+CoreFile to now also include DNS configuration for both internal and
+external access to the cluster services (the `*.apps` records).
+
+9. Once the cluster is functional, the user can look at the contents of the new
+fields `APIServerDNSConfig` and `IngressDNSConfig` in the `AWSPlatformStatus`
+field of the `Infrastructure` CR to configure their external DNS with entries
+for `api` and `*.apps` service.
+
+### API Extensions
+
+1. The AWSPlatformStatus within the PlatformStatus field of the Infrastructure
+ConfigResource (CR) is updated to contain DNS config required to configure
+the in-cluster CoreDNS solution. This same CR is available to the user post
+a successful cluster install, to configure their own DNS solution.
+
+```go
+type AWSPlatformStatus struct {
+        <snip>
+        // APIServerDNSConfig contains information to configure DNS for API Server.
+        // This field will be set only when the userConfiguredDNS feature is enabled.
+        APIServerDNSConfig []DNSConfig `json:"apiServerDNSConfig,omitempty"`
+
+        // InternalAPIServerDNSConfig contains information to configure DNS for the Internal API Server.
+        // This field will be set only when the userConfiguredDNS feature is enabled.
+        InternalAPIServerDNSConfig []DNSConfig `json:"InternalAPIServerDNSConfig,omitempty"`
+
+        // IngressDNSConfig contains information to configure DNS for cluster services.
+        // This field will be set only when the userConfiguredDNS feature is enabled.
+        IngressDNSConfig []DNSConfig `json:"ingressDNSConfig,omitempty"`
+}
+
+type DNSConfig struct {
+       RecordType  string `json:"recordType"`
+       LBIPAddress string `json:"lbIPAddress"`
+}
+
+```
+
+2. Install config is updated to allow the customer to specify if an external
+user configured DNS will be used. `UserConfiguredDNS` is added to the
+install-config and will have to be explicitly set to `Enabled` to enable this
+functionality. This config is not added to any platform specific section of
+the config because there are plans to allow this functionality in Azure and GCP
+too. The validation for this config will disallow this value being `Enabled` in
+platforms that currently do not support it.
+
+```yaml
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.9.2
+  creationTimestamp: null
+  name: installconfigs.install.openshift.io
```

Comment:
Is this a new CRD to track how the cluster was installed? Question, why would the DNS configuration not live on the infrastructure resource?

What do you see as the future of this resource?

Reply:
Yes, the DNS config is added to the AWSPlatformStatus struct in the PlatformStatus struct of the Infrastructure resource as detailed 1st in the API Extentions section.

No new CRD is being added.

This section is pointing out the addition to the install-config provided to OpenShift during cluster creation.

Reply:
Install config is a installer thing then? The installer consumes it to configure the cluster resources?

Reply:
@JoelSpeed yes, that is correct.